### PR TITLE
#12068 - Improve backend duplicate removal

### LINKED ISSUE #12068: Improve duplicate removal from merge result list [1]
### Problem Description

#12061 removed the creation date filtering logic from the backend methods to retrieve duplicate cases and contacts because it lead to filtering not working on the second case/contact of a pair. However, this required the introduction of a mechanism to manually filter out duplicate pairs after the query has been performed (because that was the exact purpose of the creation date filtering logic before). The mechanism introduced should be further improved in terms of performance and consistency as it's a) not very performant and b) leads to the result list only being half as long as the maximum amount of results specified on the merge view.

### Proposed Change

Increase the performance of the logic that removes duplicate pairs from the result, and ensure that the method is capable of returning the amount of results specified on the merge view.

### Added Value/Benefit

Better performance and consistency with the user input.

### Acceptance Criteria

- [x] The current usage of `anyMatch` has been replaced with an additional HashSet<AbstractMap.SimpleImmutableEntry<Long,Long>> that is filled while filling the resultList and then checked for duplication with a contains call
- [x] It's ensured that the result list can have as many entries as specified by the limit in the merge view AFTER the duplicates have been removed; this can probably be achieved by doubling the limit that's put into the query
- [x] Reduce the result list to the actual limit size afterwards (because there are more results returned)

### Implementation Details

_No response_

### Mockups

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by MateStrysewske**
Fixes #12068

### CODE REVIEW

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=12205

File: sormas-backend/src/main/java/de/symeda/sormas/backend/caze/CaseService.java [hunk 1]
```diff
@@ -2029,11 +2031,14 @@ public List<CaseMergeIndexDto[]> getCasesForDuplicateMerging(
 
 		TypedQuery<Object[]> typedQuery = em.createQuery(cq).setParameter("date_type", "epoch");
 		if (limit != null) {
-			typedQuery.setMaxResults(limit);
+			// Double the limit because the query result will contain each pair twice; since these duplicates will
+			// be removed, the final result list would only contain limit/2 entries otherwise
+			typedQuery.setMaxResults(limit * 2);
```
Comment (sergiupacurariu) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/caze/CaseService.java:2036]: limit set conducts to results higher than wanted

File: sormas-backend/src/main/java/de/symeda/sormas/backend/contact/ContactService.java [hunk 2]
```diff
@@ -1566,12 +1567,15 @@ public List<MergeContactIndexDto[]> getContactsForDuplicateMerging(ContactCriter
 
 		TypedQuery<Object[]> query = em.createQuery(cq).setParameter("date_type", "epoch");
 		if (limit != null) {
-			query.setMaxResults(limit);
+			// Double the limit because the query result will contain each pair twice; since these duplicates will
+			// be removed, the final result list would only contain limit/2 entries otherwise
+			query.setMaxResults(limit * 2);
```
Comment (sergiupacurariu) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/contact/ContactService.java:1572]: limit set conducts to results higher than wanted

**Review by sergiupacurariu — CHANGES_REQUESTED** [review]

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=12205

**Review by sergiupacurariu — APPROVED** [review]

**Pull request merged**